tracing-appender/src/rolling.rs

```diff
@@ -558,17 +566,90 @@ impl Inner {
                     .unwrap_or(0),
             ),
             rotation,
+            keep_last: None,
         };
         Ok((inner, writer))
     }
 
+    fn prune_old_logs(&self, keep_last: usize) {
+        let format;
+        match self.rotation {
+            Rotation::MINUTELY => {
+                format = format_description::parse("[year]-[month]-[day]-[hour]-[minute]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::HOURLY => {
+                format = format_description::parse("[year]-[month]-[day]-[hour]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::DAILY => {
+                format = format_description::parse("[year]-[month]-[day]")
+                    .expect("Unable to create a formatter; this is a bug in tracing-appender");
+            }
+            Rotation::NEVER => {
+                unreachable!("keep last N files feature is not available for `Rotation::NEVER`!")
+            }
+        }
+
+        let files = fs::read_dir(&self.log_directory).map(|dir| {
+            dir.filter_map(Result::ok)
+                .filter(|entry| {
+                    entry.file_name().to_string_lossy().contains(
+                        &self
+                            .log_filename_prefix
+                            .clone()
+                            .expect("prefix is always present"),
+                    )
+                })
+                .collect::<Vec<_>>()
+        });
+
+        match files {
+            Ok(files) => {
+                if files.len() >= keep_last {
+                    let mut sorted_files = vec![];
+                    for file in files {
+                        let filename = file.file_name().to_string_lossy().to_string();
+                        match filename.rfind('.') {
+                            Some(date_index) => {
+                                let date_str = filename[(date_index + 1)..].to_string();
+                                match Date::parse(&date_str, &format) {
+                                    Ok(date) => sorted_files.push((file.path(), date)),
+                                    Err(error) => eprintln!("error parsing the date: {error}"),
```

Comment:
This syntax doesn't compile on our current MSRV for `tracing-appender` (Rust 1.53.0): https://github.com/tokio-rs/tracing/actions/runs/2905294358/jobs/4624104893

Until the MSRV is new enough to support this syntax, we'll have to write this instead:
```suggestion
                                    Err(error) => eprintln!("error parsing the date: {}", error),
```